src/cowboy_http2.erl

```diff
@@ -227,8 +229,11 @@ loop(State=#state{parent=Parent, socket=Socket, transport=Transport,
 	receive
 		%% Socket messages.
 		{OK, Socket, Data} when OK =:= element(1, Messages) ->
-			parse(set_timeout(State), << Buffer/binary, Data/binary >>);
+			parse(set_idle_timeout(State), << Buffer/binary, Data/binary >>);
 		{Closed, Socket} when Closed =:= element(2, Messages) ->
+			%% @todo This is not an error if the client closes the
+			%% socket when Status is closing (after GOAWAY), so we
+			%% can exit with reason 'normal' in this case.
```

Comment:
I'll add a case on `http2_status` here, but `terminate/2` wraps the reason in a shutdown tuple as `exit({shutdown, Reason})`, so we don't exit with the specified Reason anyway.

Reply:
`{shutdown,...}` is a normal exit reason as far as OTP is concerned. `{shutdown, closed}` would be fine.

Reply:
That's true in the sense that no error is logged. Then we don't need to do anything here because terminate always exits with a shutdown tuple, but I can do `{shutdown, closed}` for debugging reasons.

In some situations, a process is expected to terminate with an exact reason, such when a supervisor tells a process to shutdown here: https://github.com/erlang/otp/blob/master/lib/stdlib/src/supervisor.erl#L873 or when [sys:terminate/2,3](https://erlang.org/doc/man/sys.html#terminate-2) is called with a specified Reason, that exact reason is supposed to be the exit reason.

Reply:
The comment says it's not an error *if we are already closing* and I agree. So the reason should mirror server-initiated closes, perhaps with a slight twist from `normal` to `closed` to tell us it's the client that closed and not us. So `{stop, closed, 'The client is going away.'}` or something to that effect, when we are already closing.

When we are not closing however the `socket_error` reason makes more sense because it's an unexpected close.

So depending on the `http2_status` we want one reason or the other.

I do not understand your comment about an exact reason, we are talking about the socket being closed here, there's no constraints.

Reply:
I'll do as you suggested.

Just to explain the previous comment: There are no constraints in the socket closed case, but the comment was generally about wrapping any reason in `{shutdown, ...}`. For unexpected errors, shouldn't an error be logged by OTP, if the process is supposed to behave like OTP processes do? (I'm not saying an unexpected socket close is an unexpected error; it's quite normal.)

Since ranch is used and not the OTP supervisor, the point about supervisor shutdown isn't valid. The point about `sys:terminate/2,3` is, but it's a rather unimportant point. If the caller calls `sys:terminate/2,3` with an abnormal reason, I guess an error is expected to be logged.

Reply:
Ranch can shut down via `shutdown` just like the OTP supervisor.

When we crash there are appropriate error events. When the user code crash we produce easier to read error reports.

Reply:
Yeah, the difference is that `ranch_conns_sup` doesn't care about the Reason in the `{'DOWN', _, process, Pid, _}` message while `supervisor` requires the Reason to be the atom `shutdown` or otherwise it's regarded it as an error which is logged. (Now, I don't think anyone wants to use the regular OTP `supervisor` with a Cowboy connection process.)

I'm sure the error handling/reporting/logging is appropriate as it is. Cowboy's exit reasons are surely more readable when debugging.